### PR TITLE
Implement wc(s)width and ABMON_1..12 in nl_langinfo

### LINKED ISSUE #500: Strange nl_langinfo errors
When running `ls -la`, the following is printed

```
mlibc: nl_langinfo item 393230 is not implemented properly
mlibc: nl_langinfo item 393231 is not implemented properly
mlibc: nl_langinfo item 393232 is not implemented properly
mlibc: nl_langinfo item 393233 is not implemented properly
mlibc: nl_langinfo item 393234 is not implemented properly
mlibc: nl_langinfo item 393235 is not implemented properly
mlibc: nl_langinfo item 393236 is not implemented properly
mlibc: nl_langinfo item 393237 is not implemented properly
mlibc: nl_langinfo item 393238 is not implemented properly
mlibc: nl_langinfo item 393239 is not implemented properly
mlibc: nl_langinfo item 393240 is not implemented properly
mlibc: nl_langinfo item 393241 is not implemented properly
mlibc: nl_langinfo item 393230 is not implemented properly
mlibc: nl_langinfo item 393231 is not implemented properly
mlibc: nl_langinfo item 393232 is not implemented properly
mlibc: nl_langinfo item 393233 is not implemented properly
mlibc: nl_langinfo item 393234 is not implemented properly
mlibc: nl_langinfo item 393235 is not implemented properly
mlibc: nl_langinfo item 393236 is not implemented properly
mlibc: nl_langinfo item 393237 is not implemented properly
mlibc: nl_langinfo item 393238 is not implemented properly
mlibc: nl_langinfo item 393239 is not implemented properly
mlibc: nl_langinfo item 393240 is not implemented properly
mlibc: nl_langinfo item 393241 is not implemented properly
```

Seen during testing for the mlibc on Linux project.

### PR DESCRIPTION
**Pull request by no92**
Fixes #500.

### CODE REVIEW

**Comment by avdgrinten** [comment]
The wcwidth implementation is really ugly (the code table will have to be changed when the Unicode standard is updated). While this is not a problem for wcwidth itself, there are lots of other functions (e.g., `iswalpha()`) that we can't implement in the same style. It would be better to write a script that uses ICU to extract the required data.

**Comment by avdgrinten** [comment]
Actually, ICU does *not* implement the equivalent of this function so maybe we should just keep it as it is (?).

**Comment by 64** [comment]
This should probably have a test, since it's new functionality.

**Comment by no92** [comment]
Added test and rebased against master.

File: tests/posix/wcwidth.c [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	tests_run++;
+	int actual_width = wcwidth(c);
+	if (actual_width != expected_width) {
+		fprintf(stderr, "ERROR: wcwidth(U+%04x) returned %d, expected %d\n", c, actual_width, expected_width);
+		test_failures++;
+	}
+}
+
+int main() {
+	assertWidthIs(1, 'a');
```
Comment (ArsenArsen) [inline tests/posix/wcwidth.c:34]: Does this succeed on host by setting locale before the line above?

Comment (no92) [reply tests/posix/wcwidth.c:36]: Setting locale makes it work, other than one test excluded from host testing.

File: tests/posix/wcwidth.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	assertWidthIs(2, L'Ｂ');
+	assertWidthIs(2, L'Ｃ');
+	assertWidthIs(2, L'中');
+	assertWidthIs(2, L'文');
+	assertWidthIs(2, 0x679C);
+	assertWidthIs(2, 0x679D);
+	assertWidthIs(2, 0x2070E);
+	assertWidthIs(2, 0x20731);
+
+#ifdef HOST_LIBC
```
Comment (64) [inline tests/posix/wcwidth.c:50]: This should be `USE_HOST_LIBC`

**Review by 64 — CHANGES_REQUESTED** [review]

**Review by 64 — APPROVED** [review]

File: tests/posix/wcwidth.c (hunk 2 above)
Comment (no92) [reply tests/posix/wcwidth.c]: Changing this to `USE_HOST_LIBC` leads to CI failures, though.

Comment (no92) [reply tests/posix/wcwidth.c]: Nevermind, negating this does wonders.

File: options/ansi/generic/wchar-stubs.cpp [hunk 3]
```diff
@@ -228,17 +228,528 @@ size_t wcsnrtombs(char *mbs, const wchar_t **wcsp, size_t wc_limit, size_t mb_li
...
+		ucs == 0x2028 || ucs == 0x2029 || (0x202A <= ucs && ucs <= 0x202E) ||
+		(0x2060 <= ucs && ucs <= 0x2063)) {
+		return 0;
+	}
+
+	// C0/C1 control characters.
+	if (ucs < 32 || (0x07F <= ucs && ucs < 0x0A0)) return -1;
+
+	// Combining characters with zero width.
+	if (intable(ZERO_WIDTH, sizeof(ZERO_WIDTH)/sizeof(struct width_interval), ucs)) return 0;
```
Comment (64) [inline options/ansi/generic/wchar-stubs.cpp:738]: The spacing should be fixed here.

**Review by 64 — APPROVED** [review]

**Pull request merged**